### PR TITLE
feat: Allows buying a domain from the marketplace contract #34

### LINKED ISSUE #14: List domain: Hooking up controller with RNS Smart contracts
Hooking it up with the smart contracts (dependency smart contracts done) for the List for sale functionality.

**User Story:** [List Domain for Sale](https://github.com/rsksmart/rif-marketplace-project/issues/11)

### PR DESCRIPTION
**Pull request by jurajpiar**
Closes #14.

**EDIT: The tutorial has been replaced with https://hackmd.io/CEIhzk6ZTwSOEzLIyToLsQ?view .** Keeping the current one for reference. (edited by @vojtechsimetka )

# Dependencies:
0. NVM
1. Docker
2. the [Dev](https://github.com/rsksmart/rif-marketplace-dev) project for database, db admin and the ganache network
3. the [Rns-Dev-Env](https://github.com/rsksmart/rif-marketplace-rns-dev-env) project to deploy the rns suite to ganache (You'll need to `npm i` in node 10 (nvm use 10))
4. the [Cache](https://github.com/rsksmart/rif-marketplace-cache) project's **feat/rns-entities** branch for rns cache

# Setup _(After installing all required npm modules)_:
1. run Dev with `docker-compose-up`
2. copy the first **private key** from the output and add account to your wallet
3. run the deploy script with `npx truffle deploy --network ganache`
4. copy the Rif token address from the output and add token to your wallet account
5. copy Rif token, RSK Owner and Simple Placements addresses from the output of _(3.)_ to .env file in the UI project
6. in the Cache project add _config/local.json_ (you can ignore the addresses. the BC part of the Cache doesn't work. this is just to prevent it from error-ing out):
```
{
    "host": "localhost",
    "port": 3030,
    "blockchain": {
        "provider": "ws://127.0.0.1:4445/websocket"
    },
    "storage": {
        "contractAddress": "0x83C5541A6c8D2dBAD642f385d8d06Ca9B6C731ee",
        "events": [
            "CapacitySet",
            "MaximumDurationSet",
            "PriceSet"
        ],
        "eventsEmitter": {
            "confirmations": 1,
            "startingBlock": "genesis"
        }
    },
    "rns": {
        "contractAddress": "0xd6681FD2d471c044BDdDc9DF00D1Af195Bd58bfa",
        "events": [
            "Transfer",
            "ExpirationChanged"
        ],
        "eventsEmitter": {
            "confirmations": 1,
            "startingBlock": "genesis"
        },
        "newBlockEmitter": {
            "polling": false
        }
    },
    "rns-reverse": {
        "contractAddress": "0x8650B271F7cC565aa604F9f98Aa1A5B723B5b2f3",
        "events": [
            "NameChanged"
        ],
        "eventsEmitter": {
            "confirmations": 1,
            "startingBlock": "genesis"
        },
        "newBlockEmitter": {
            "polling": false
        }
    },
    "db": "postgres://dev:dev@localhost/rif_marketplace_cache",
    "log": {
        "level": "debug",
        "path": "./logs"
    }
}
```
7. comment out lines 15-19 in the src/rns/hooks/domain-offer.hook.ts
8. In the browser navigate to `http://localhost:8080`, where
```
db=rif_marketplace_cache
name=dev
pass=dev
```
9. insert into rns_domain table
```
ownerAddress=your new wallet account address
name=bob.rsk _it is important that you use [bob, alice, charlie] for now as these are def owned by the first account_
```
10. run UI
11. follow the User story [#11 (List RNS Domain for Sale)](https://github.com/rsksmart/rif-marketplace-project/issues/11)
12. copy the second **private key** from the output and add account to your wallet
13. insert into rns_domain_offers table
```
sellerDomain=same as in rn_domain.name
rest can be anything
```
14. follow the User story [#9 (Buy an RNS Domain)](https://github.com/rsksmart/rif-marketplace-project/issues/9)

### CODE REVIEW

File: src/components/organisms/Header.tsx [hunk 1]
```diff
@@ -1,26 +1,13 @@
...
-      <Account
-        web3={web3}
-        networkName={networkName}
-        account={account}
-        setProvider={setProvider}
-      />
-    )}
-  </Web3Provider.Consumer>
-);
+import Login from 'components/atoms/Login';
```
Comment (farinalvaro) [inline src/components/organisms/Header.tsx:6]: Love this extraction, will take note to move it to RIF-UI later 🙂

File: src/ui-config.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (vojtechsimetka) [inline src/ui-config.json:1]: Please remove the corresponding `.env` values and ideally introduce `.env` value called network set to `ganache`

File: .env [hunk 3]
```diff
@@ -1,2 +1,2 @@
 REACT_APP_CACHE_ADDR=http://localhost:3030
-REACT_APP_LOG_LEVEL=error
\ No newline at end of file
+REACT_APP_LOG_LEVEL=debug
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply .env:2]: uhoh...yea my bad. forgot to remove that.

[the remark this answers is not in the document]
Comment (jurajpiar) [reply .env:2]: I'll use .local.env in the future. thanks

File: src/components/templates/MarketPageTemplate.tsx [hunk 4]
```diff
@@ -11,6 +11,7 @@ export interface MarketPageTemplateProps {
   filterItems: React.ReactNode;
   headers: TableHeaders;
   itemCollection: MarketItemType[];
+  accountRequired?: boolean;
```
Comment (farinalvaro) [inline src/components/templates/MarketPageTemplate.tsx:14]: This is great, intuitive and simple!

File: src/components/pages/rns/buy/DomainOffersCheckoutPage.tsx [hunk 5]
```diff
@@ -103,19 +125,51 @@ const DomainOffersCheckoutPage = () => {
...
+                    isProcessing: true,
+                }
+            })
+
+            const marketPlaceContract = await Contract(ERC721SimplePlacements).at(marketPlaceAddress)
+            const rifContract = await Contract(ERC677).at(rifTokenAddress)
+            const rnsContract = await Contract(ERC721).at(rnsAddress)
+
+            const owner = await rnsContract.ownerOf(tokenId)
+            console.log(`ownerOf${domainName}:${owner}`)
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/pages/rns/buy/DomainOffersCheckoutPage.tsx:146]: Removed that log in the previous PR so should be gone when rebased. Don't need to log that actually. Just a left over.

File: src/ui-config.json (hunk 2 above)
Comment (jurajpiar) [reply src/ui-config.json:1]: done

**Pull request merged**